Add queries from external apis (resolves #2)

anttikup opened the pull request:
Adds 
- common format for external api handlers
    - an example handler
- looking up of handlers per language
- merging of results from reqular queries and these into one result set

Closes #2

pennane commented on tests/wikidata/test_queries.py:
```diff
@@ -26,3 +27,9 @@ def test_bind_sparql_query():
 
     with pytest.raises(Exception):
         bind_sparql_query("{{x}}", x='1', y="2")
+
+    with pytest.raises(NotImplementedError):
+        bind_sparql_query("{{x}}", x=[1, 2, 3])
+
+    with pytest.raises(NotImplementedError):
+        bind_sparql_query("{{x}}", x=["a", "b", 3])
```
could be beneficial to have additional context for the tests in the shape of comments. Or if pytest implements some kind of labels, then those could be used.

Now it is a bit ambiguous on what is unimplemented

anttikup replied:
Thanks! Refactored the tests into separate functions to and the code to distinct possible, but not implemented, inputs and erroneous inputs.

pennane commented on tests/wikidata/test_api.py:
```diff
@@ -27,3 +28,11 @@ def test_build_composite_description():
         == 'kissa (kotikissa, kesykissa) [= cat]: an animal'
+
+
+def test_yield_external_results():
+    searched_word = StrInLanguage('eurooppalainen lyhytkarva', 'fi')
+    results = list(yield_external_results(searched_word))
+
+    assert len(results) == 1
+    assert results[0].id == 'Q20793'
```
I would extract the `Q20793` to some constant variable

anttikup replied:
I put the value in a constant and tried to give it a better name, although I'm not sure what to call it since it is not very important what the page is about.

pennane commented on suggestedimages/wikidata/queries/util.py:
```diff
@@ -37,13 +37,19 @@ def bind_sparql_query(query, **params):
-        else:
-            raise Exception(f'Invalid type: {str(value)}')
+        query = replace_placeholders(query, key, value)
 
     return query
+
+
+def replace_placeholders(query, key, value):
+    if type(value) == str:
+        return query.replace("{{" + key + "}}", f'"{value}"')
+    elif type(value) in [int, float]:
+        return query.replace("{{" + key + "}}", str(value))
+    elif type(value) == Identifier:
+        return query.replace("{{" + key + "}}", str(value))
+    elif type(value) == list and all(type(member) == str for member in value):
+        return query.replace("{{" + key + "}}", " ".join(f'"{member}"' for member in value))
+    else:
+        raise NotImplementedError(f'Type: {type(value)}')
```
👍

pennane commented on suggestedimages/wikidata/queries/__init__.py:
```diff
@@ -1,20 +1,43 @@
+
+
+def property_has_any_of_values(property: str, values: list[str], limit=50):
+    """Return items where property `property` has any of values `values`.
+    """
```
This comment might be redundant. The function name is already as (if not more) clear on what the purpose of it is.

anttikup replied:
:+1:

pennane commented on suggestedimages/external/__init__.py:
```diff
@@ -0,0 +1,11 @@
+from .yso import YSO
+
+
+by_language = {
+    'fi': [
+        YSO('fi')
+    ]
+}
```
What by language?

`something_by_language` could offer more context

anttikup replied:
Yes, you're right. I renamed it and the whole module.

pennane commented on suggestedimages/external/yso.py:
```diff
@@ -0,0 +1,40 @@
+                [item['altLabel']] if 'altLabel' in item else []
+            ) \
+            for item in json_response['results'] \
+            # We're only interested in exact maches
```
why? The code already tells how

pennane commented on suggestedimages/wikidata/queries/__init__.py:
I am not 100% sure on the python best practices, but I believe that the `__init__.py` files should not contain logic inside of them and should only be used for initialization of the package.

I would move the functions defined in the `__init__.py` into some new submodules.

anttikup replied:
Yes, you're probably right. Fixed.

pennane reviewed (commented):
Nice! :shipit:  Structure looks good and easy to follow.

Commits are descriptive (although maybe there are too many of them :D).

Left quite a few code-level comments. Some of them might not make sense nor follow python best practices.

anttikup commented:
Thanks, for the good review!

> The project is well structured. Naming is mostly good, though some class names like "FINTO" and "YSO" (which I guess refer to external services) were not immediately apparent, maybe a comment on the purpose of the services could come in handy. 

I added links to the services in the files and also a description in the external_apis directory.

> The help link in the local version of the app doesn't seem to return anything.

Added a rough help text in English for now, although this should be locale dependent.

> Some commit comments could be more descriptive like "Added new type, list of strings and refactored for clarity", in this context "new type" doesn't necessary mean much.

Good point, thanks! I'll try to add more context.

> The description in the README is okay, but could maybe use a bit more detail.
> Instructions could use some work, I'm think you need to create a venv before you can activate it after cloning the project? I also think that the command to activate venv is platform dependant.

Added more details. I don't really know how to activate it in other platforms.
